packages/babel-parser/src/parse-error.js

```diff
@@ -0,0 +1,211 @@
+// @flow
+
+import { Position } from "./util/location";
+import type { NodeBase } from "./types";
+import {
+  type ParseErrorCode,
+  ParseErrorCodes,
+  type ParseErrorCredentials,
+} from "./parse-error/credentials";
+
+const ArrayIsArray = Array.isArray;
+const {
+  assign: ObjectAssign,
+  defineProperty: ObjectDefineProperty,
+  getPrototypeOf: ObjectGetPrototypeOf,
+  keys: ObjectKeys,
+} = Object;
+
+type ToMessage<ErrorDetails> = (self: ErrorDetails) => string;
+
+const StandardMessage = Symbol("StandardMessage");
+
+// This should really be an abstract class, but that concept doesn't exist in
+// Flow, outside of just creating an interface, but then you can't specify that
+// it's a subclass of `SyntaxError`. You could do something like:
+//
+// interface IParseError<ErrorDetails> { ... }
+// type ParseError<ErrorDetails> = SyntaxError & IParseError<ErrorDetails>;
+//
+// But this is just a more complicated way of getting the same behavior, with
+// the added clumsiness of not being able to extends ParseError directly. So,
+// to keep things simple and prepare for a Typescript future, we just make it a
+// "private" superclass that we exclusively subclass:
+
+export class ParseError<ErrorDetails> extends SyntaxError {
+  code: ParseErrorCode;
+  reasonCode: string;
+
+  loc: Position;
+  details: ErrorDetails;
+
+  // There are no optional fields in classes in Flow, so we can't list these
+  // here: https://github.com/facebook/flow/issues/2859
+  // syntaxPlugin?: SyntaxPlugin
+  // missingPlugin?: string[]
+
+  constructor({
+    loc,
+    details,
+  }: {
+    loc: Position,
+    details: ErrorDetails,
+  }): ParseError<ErrorDetails> {
+    super();
+
+    this.loc = loc;
+
+    ObjectDefineProperty(this, "details", {
+      value: details,
+      enumerable: false,
+    });
+
+    // $FlowIgnore
+    if (details.missingPlugin) {
+      ObjectDefineProperty(this, "missingPlugin", {
+        get() {
+          return this.details.missingPlugin;
+        },
+        enumerable: true,
+      });
+    }
+
+    return this;
+  }
+
+  clone({
+    loc,
+    details,
+  }: {
+    loc?: Position,
+    details?: ErrorDetails,
+  } = {}) {
+    return new (ObjectGetPrototypeOf(this).constructor)({
+      loc: loc || this.loc,
+      details: { ...this.details, ...details },
+    });
+  }
+
+  get pos() {
+    return this.loc.index;
+  }
+}
+
+function toParseErrorClass<ErrorDetails>(
+  toMessage: ToMessage<ErrorDetails>,
+  credentials: ParseErrorCredentials,
+): Class<ParseError<ErrorDetails>> {
+  return class extends ParseError<ErrorDetails> {
```

Comment:
If I'm reading this file correctly, the new approach creates a new class for every possible error. Would it be possible to use a single class instead (`ParseError`), and pass the error code or the `toMessage` function as a constructor argument?

Reply:
Short answer is not trivially, but there are kind of two categories of reasons for this, "design reasons" (legitimate), and "Flow reasons" (unfortunate):

1. To approach this from the "design" side first, this is in many senses fundamentally asking whether Errors should be nominal types or not. If we were to try to collapse all errors employ "one re-usable class", then you can't robustly express something like:

```javascript
export type DeferredStrictErrorClass =
  | typeof Errors.StrictNumericEscape
  | typeof Errors.StrictOctalLiteral;
```

This is because, despite these errors having different names, they all fundamentally have the same "shape", which is just `ParseError<{||}>`. That is to say, they're just type aliases, which can cross-satisfy. So ParseErrorA *can* satisfy ParseErrorB, as long as they have identical `ErrorDetails`, which is at it turns out is very common because the vast majority of ParseError types are implicitly parameterized with empty details (`{||}`). This is made even worse by the fact that, from the user's perspective (the user being us in this case), it seems arbitrary which types are nominal and which aren't. Because, if the ParseError *does* have a unique shape, all of a sudden it behaves effectively like a nominal type. For example, no other error is ever going to have a `surrogateCharCode` member, so for all practical purposes, `Errors.ModuleExportNameHasLoneSurrogate` *is* a nominal type, and you can safely write a function like this:

```function doSomethingWithALoneSurrogateError(error: Errors.ModuleExportNameHasLoneSurrogate) {
// Can't pass just any old error here!
}
```

So by collapsing everything to one class, you get this weird world that can lull you into thinking you can use the type system to very specifically type certain functions, only to be surprised that that's not the case... but only sometimes.

Now, in Typescript (and other type systems), there is a tricky solution to this. You can force nominality with phantom fields that are only present at type-check time. In my original Typescript implementation, I actually wrote it the way you proposed, and employed this trick, something like this:

```
type ParseErrorClasses<T extends AllErrorDetails> = {
  [ReasonCode in keyof AllErrorDetails]:
    ParseErrorClass<AllErrorDetails[ReasonCode]> & { NominalType: ReasonCode };
};
```

As you can see here, we're actually using the `ReasonCode` to further narrow the acceptable type (without making it a static field, which has it's own issues). In this scenario, even if ErrorDetailsA === ErrorDetailsB, Error.A !== Error.B, because `{ NominalType: "A" }` does *not* satisfy `{ NominalType: "B" }`. Now, this is unfortunately not doable in Flow, although even if we did go this route in Typescript it has it's own downsides. But, to summarize the "design decision", making each error a separate class allows us to be more expressive with our types, which I do make use of throughout the code.

2. To just address a further Flow-specific thing, Flow is very finicky about shuttling parameter arguments around. Having a specific class-per-error allows Flow to keep the different parameterized types separate. This is due to a know bug (previously "expected behavior") where Flow "pushes" function unions down to their parameters. Here is a reproduction: https://github.com/facebook/flow/issues/8838. I actually got really close to doing something like you're describing, but then Flow says "oh wait a minute, shouldn't I just merge *all* the constructor calls to be unions of each-other then?" So basically instead of constructor<T>(T) | constructor<U>(U), it transforms it into constructor<T | U>({ ...T, ...U }), meaning every error will accept the arguments for every error defined anywhere (ahhhhhhh!) What makes this bug even more insidious is that it only happens on *calling*. That is to say, everything *seems* fine until you actually raise the error. This was super frustrating because I was getting this all working in the try flow playground, and I would try raising individual errors, and everything would be fine, but then I'd try `make flow` and it would be broken. Finally, I tried calling two *separate* raises and it broke in the playground.  Anyways, if and when Flow fixes this (they *do* want it to work like Typescript, but its part of a year-long rewrite), then *that* part of the problem would be solved. So, when we move to Typescript, something like this *would* be easier.

3. Just to add one more consideration, I tried to do my best to not make any internal functionality public, such that any "abuses" of the API would be fine. So for example:

```javascript
// We don't want this:
const ParseErrorClass = result.errors[0].constructor;
const trickyError = new ParseErrorClass(myOwnToMessage);
// But we're fine with this:
const sureDoThis = new ParseErrorClass({ loc: new Position() });
```

Reply:
Since all the errors have a distinct `reasonCode` property, doesn't it make them nominal?

Reply:
Right, but I have been incapable of getting this to work in Flow, specifically, when first routed through `toParseErrorClasses` *and* in `raise`. It's "easy" to make `toParseErrorClass` (not-plural) work, it totally breaks down for `toParseErrorClasses` (do to parameter-type extraction not working, and $ObjiMap not working, and type-parameter type extraction not working, any of which would provide an escape hatch, all of which will "in theory" work when that bug I linked to is fixed).

To briefly summarize, unlike in Typescript, I don't have access to `ReasonCode` at the right time, since it's "trapped" in the keys of the definition object. In Typescript, this is irrelevant, since `keyof` *actually works*, and this is trivial:
```javascript
toParseErrorClasses<T>(errors: T):
  { [ReasonCode keyof T]: ParseError<ReasonCode, Parameters<T[ReasonCode]>>[0] }
```
Flow however has no way of extracting the parameters of types from a function (well, it almost does, but it's broken, you can find several issues on it). Hence the need to bounce everything through that annoying `_` function, which means we don't know the `ReasonCode` at the point of creating our `ParseError` class. So, not a problem you might say, we'll just create an intermediate "shuttle" type to "carry" the `ErrorDetails` to the next step (which we kind of also do in the current implementation), but then you run into Flow's inability to properly extract (infer) the parameterized types. I will try to find my saved tryFlow examples, but even if we dramatically loosen the constraints around the "usability" of the API and start forcing the user to type more and more boilerplate, you then run into the bug I linked to above, where without the "class separation" to artificially prevent Flow from getting confused and thinking these are all overloads of the same constructor/raise function and thus union-ing the parameter types. 

Anyways, I definitely would prefer something *like* this (I still think there are some issues with a method like this, in particular I find it weird to have to pass in a reasonCode that it already knows ahead of time what it should be exactly, and I also dislike that if a user gets ahold of these classes, there's no type system to enforce that -- but that's a different matter to discuss if we can first get this working at all). In fact, if you recall this is what originally made me angry enough to try to *first* port the entire parser to typescript just to avoid having to find workarounds to well-known bugs in Flow. That being said, I can't tell from this whether you tried it already, but if you have and you've gotten it working with `toParseErrorClasses` and `raise`, I'd actually be super open to looking into it further, since I do agree that the current internals aren't ideal, but as far as I can tell provide the closest thing to the ideal API we're going to get before switching to Typescript.

Reply:
I tried this: https://github.com/tolmasky/babel/pull/1

Reply:
Right, but the only purpose for `toParseErrorClasses` is to avoid this boilerplate and duplication ;) If you pass in all the info to `_` yourself, not only does it work (since we already know we can make `x = toParseErrorClass<"blah">()` work), it now actually makes `toParseErrorClasses` redundant, since you can remove it and get the same behavior as just manually calling `toParseErrorClass` on each set of infos yourself manually:

```javascript
const Errors = toParseErrorClasses(_ => ({
 Error1 = _<"Error1>("message"),
 Error2 = _<"Error2", {| x: string |}>(({ x: string }) => `...`)
});
```

Doesn't provide any "type inference magic" over the following:

```javascript
const Errors = {
 Error1 = toParseErrorClass("Error1",  "message"),
 Error2 = toParseErrorClass<{| x: string |}>("Error2",  ({ x: string }) => `...`)
};
```
It only saves you from typing the long function name `toParseErrorClass`, which is equivalently resolved by just renaming it to that manually:

```javascript
import { toParseErrorClass as _ } from "./parse-error";

const Errors = {
 Error1 = _<"Error1">( "message"),
 Error2 = _<"Error2">({ x: string }) => `...`)
};
```

But in all cases, we now have to specifically, repeatedly, provide all the information at the call-site, which skirts around the original goal of avoiding the change for errors through boilerplate, and the noisiness of having the error definitions dominated by the same error codes and type info over and over. Unfortunately, the need for the `_` in Flow makes this a little less clear, since in Typescript the following is so much cleaner:

```javascript
Errors = toParseErrorClasses({
 Error1: "message",
 Error2: ({ x: string }) => `...`) // no need even for <{ x: string }>!
});
```

Reply:
Alright, I just pushed an update that seems to work. I'll describe it here tomorrow, but the most important point is that this version doesn't have any sort of `ParseError` class at all, so we're I think just about as close as possible to introducing no new outside-observable changes. Still want to clean it up a bit more and then explain the change here, but will get to it tomorrow.

Reply:
Meant to put this here originally, just copying it here in case someone else is reading:

OK, I still want to clean up a few things, and add more comments, etc., but here is an explanation of the current code. First, the starting with the goals:

1. First off, we want to have the named and type-checked parameters of course.
2. We'd like to have an easy way of defining these errors.
3. We'd like for these changes to minimally affect the publicly exposed API.

As mentioned previously, all of these are trivially achievable in Typescript, thanks to much better inference and way better functioning object-mapping support. That being said, the original implementation of this PR didn't satisfy (3), due to creating a class per error (which I think might have some potentially benefits as an *API choice* in the future, but I agree that we should avoid it now if possible, since it is a *separate* concern from the goals above, and may tie us to a decision we need to support in the future). 

I did manage to find a workaround to the bug that was preventing us from having the nominal types without a subclass, using an indirection trick (if you do `[V][0]`, Flow stops "merging" all the types together). Unfortunately, this only works in later versions of Flow, and upgrading even a few versions up puts us in the new [type-first](https://flow.org/en/docs/lang/types-first/) mode in Flow, which throws thousands of errors. So, didn't seem worth it to try to block everything on that big update.

However, as I was trying to remove the subclasses-per-error-type stuff, I realized that this is still technically a problem even if we do get it down to *one* combined `ParseError` class, as that still deviates from the current behavior of just using `SyntaxError`'s. Given that the current code already did some fancy stuff to hide that fact (like overriding the constructor name), it seems that whether we add 1 class to the hierarchy or 2 doesn't make that much of a difference (unless you're doing really weird stuff like `error1 instanceof error2.constructor` or something), and by that I mean they're both "bad", not that they're both "good" so we shouldn't worry about it. Additionally, I noticed that in some of the most critical places where we used our nominal types, the original PR wasn't taking advantage of the nominality (such as with `raise`).

So, the new approach works much harder on fixing (2) completely (by going back to a "just SyntaxErrors with extra properties" approach), and keeping it "as nominal" as the original PR (which is to say, unfortunately not nominal enough, but it's really only affecting the quality of the error messages currently), and putting us in a place that definitely has identical "value structure" to what we'd get in a future move to Typescript, with *almost* most of the type conveniences.